src/ast.rs

```diff
@@ -68,39 +65,202 @@ impl<'a> Program<'a> {
-        // on a data block. For that, we most likely need to use something like skip lists.
-        for (index, block) in blocks.iter_mut().enumerate() {
+        // on a data block. Instead, we replace them with REM.
+        // 3. Flatten FOR blocks and generate equivalent code without FOR.
+        for block in blocks {
             match block {
                 Block::Line {
                     line_number,
-                    ref mut statement,
+                    statement_source,
+                    statement,
                     ..
                 } => {
                     match statement {
-                        Statement::Data(ref mut statement_data) => {
-                            data.append(statement_data);
+                        Statement::Data(mut statement_data) => {
+                            self.data.append(&mut statement_data);
+                            self.blocks.push(Block::Line {
+                                line_number,
+                                statement_source,
+                                statement: Statement::Rem,
+                            })
                         }
-                        _ => (),
+                        other_statement => self.blocks.push(Block::Line {
+                            line_number,
+                            statement_source,
+                            statement: other_statement,
+                        }),
                     };
-                    if block_index.insert(*line_number, index).is_some() {
-                        return Err(Error::DuplicateLineNumber {
-                            line_number: *line_number,
+                    self.index_last_block()?;
+                }
+                // Flatten FOR...NEXT by replacing FOR statement with LET preamble and
+                // IF statement, NEXT statement with GOSUB.
+                Block::For {
+                    for_line,
+                    blocks,
+                    next_line,
+                } => {
+                    // check that control variables match
+                    if for_line.for_statement.control_variable
+                        != next_line.next_statement.control_variable
+                    {
+                        return Err(Error::InvalidControlVariable {
+                            src_line_number: for_line.line_number,
+                            control_variable: format!(
+                                "{}",
+                                for_line.for_statement.control_variable
+                            ),
                         });
                     }
+
+                    // LET .. = limit
+                    let limit = NumericVariable::Limit {
+                        line_number: for_line.line_number,
+                    };
+                    self.blocks.push(Block::Line {
+                        line_number: internal_line_numer,
+                        statement_source: Span::new(CompleteStr("")),
```

Comment:
This is kind of misleading. What do you think about using an optional for the `statement_source` or a `Block::Phony`?

Reply:
Ah, this is good proposal. Will do it in sub sequential review.